oak_channel/SPEC.md

```diff
@@ -43,244 +42,157 @@ The service MUST NOT send request messages.
+The invocation ID and message length is already encoded into the frame header,
+so an encoded message does not have a header of its own and MUST only consist of
+a body containing the encoded bytes of the request or response:
 
 ```text
  0                   1                   2                   3
  0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
 +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|                             length                            |
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|                         invocation_id                         |
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 |                                                               |
 +                         body bytes...                         +
 |                                                               |
 +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+ ...
 ```
 
-<!-- Diagram generated with https://www.luismg.com/protocol/, using the schema
-"length:32,invocation_id:32,body bytes...:64"  -->
+<!-- Diagram generated with https://www.luismg.com/protocol/, using the spec
+"body bytes...:64"  -->
 
-- `length`, u32, little endian
-
-  Length of the entire message, including this field.
-
-- `invocation_id`, u32, little endian
-
-  Identifies the invocation the request initiates.
+### Frame Layer
 
-- `body`, variable length byte array
+Messages are sent and received as frames with a maximum total length of 4,096
+bytes per frame. Each message is fragmented into a set of one or more frames.
 
-  Byte encoded method parameters.
+#### Frame
 
-#### Response Message
+MUST consist of a header and a body. The encoded header is 16 bytes long and
+MUST consist of the following fields:
 
-A byte encoded response message MUST consist of the following fields:
+- `protocol_version`, unsigned 16-bit little-endian integer
 
-```text
- 0                   1                   2                   3
- 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|                             length                            |
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|                         invocation_id                         |
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|                                                               |
-+                         body bytes...                         +
-|                                                               |
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+ ...
-```
+  The version of the communication protocol in use. This value MUST be 1 for
+  this version of the specification.
 
-<!-- Diagram generated with https://www.luismg.com/protocol/, using the schema
-"length:32,invocation_id:32,body bytes...:64" -->
+- `body_length`, unsigned 16-bit little-endian integer
```

Comment:
Wouldn't it be better to have the total frame length here instead of just the body? That way, if we evolve the protocol to a future version, as long as we keep these two fields the same, a parser would be able to at least delimit the entire frame, even if it cannot parse it. At the moment, the parser would have to know the exact length of all the fields for the specific version. Also, even if we have only one version, it seems nice to logically separate the delimitation of frames via their length from their actual parsing. WDYT?